### PR TITLE
Informix r3

### PR DESCRIPTION
**Pull request by vivek-22**
Why
The release 3 items for Informix DB sensor needs changes

What
Changes done for release 3 items for informix DB sensor. The following new metrics has been added as a part of the release 3.
1)db.seq.scan.count
2)db.seq.scan.table.count
3)db.overflow.lock.count
4)db.overflow.transaction.count
5)db.overflow.user.count

### CODE REVIEW

File: rdb/src/main/java/com/instana/dc/rdb/impl/informix/InformixDc.java [hunk 1]
```diff
@@ -110,6 +111,17 @@ private void registerMetricsMetadata() {
                 new MetricDataConfig(InformixUtil.DB_DATABASE_NLS_ENABLED_SQL, MetricCollectionMode.SQL, List.class, DB_DATABASE_NLS_ENABLED_KEY));
         MetricsDataConfigRegister.subscribeMetricDataConfig(DB_DATABASE_CASE_INCENSITIVE_NAME,
                 new MetricDataConfig(InformixUtil.DB_DATABASE_CASE_INCENSITIVE_SQL, MetricCollectionMode.SQL, List.class, DB_DATABASE_CASE_INCENSITIVE_KEY));
+        MetricsDataConfigRegister.subscribeMetricDataConfig(DATABASE_LOCK_TABLE_OVERFLOW_NAME,
+                new MetricDataConfig(InformixUtil.LOCK_OVF_SQL, MetricCollectionMode.SQL, Number.class));
+        MetricsDataConfigRegister.subscribeMetricDataConfig(DATABASE_TRANSACTION_OVERFLOW_NAME,
+                new MetricDataConfig(InformixUtil.TRANSACTION_OVF_SQL, MetricCollectionMode.SQL, Number.class));
+        MetricsDataConfigRegister.subscribeMetricDataConfig(DATABASE_USER_OVERFLOW_NAME,
+                new MetricDataConfig(InformixUtil.USER_OVF_SQL, MetricCollectionMode.SQL, Number.class));
+
```
Comment (debnathbikram) [inline rdb/src/main/java/com/instana/dc/rdb/impl/informix/InformixDc.java:120]: remove this space

File: rdb/src/main/java/com/instana/dc/rdb/impl/informix/InformixDc.java [hunk 2]
```diff
@@ -193,11 +208,14 @@ private void scheduleCustomPollRate(int pollInterval, IntervalType intervalType)
     private void parseCustomAttributes(Map<String, Object> properties) {
         Map<String, Object> customInput = (Map<String, Object>) properties.get("custom.input");
         String[] dbNames = ((String) customInput.get("db.names")).split(",");
+        int sequentialScanCount = (Integer)customInput.get("db.sequential.scan.count");
         StringBuilder sb = new StringBuilder(Constants.SINGLE_QUOTES + dbNames[0] + Constants.SINGLE_QUOTES);
         for (int i = 1; i < dbNames.length; i++) {
             sb.append(Constants.COMMA).append(Constants.SINGLE_QUOTES).append(dbNames[i].trim()).append(Constants.SINGLE_QUOTES);
         }
         tableSpaceSizeQuery = String.format(InformixUtil.TABLESPACE_SIZE_SQL, sb);
+        sequentialScanQuery =  String.format(InformixUtil.DB_SEQ_SCAN_SQL,sb,sequentialScanCount);
```
Comment (debnathbikram) [inline rdb/src/main/java/com/instana/dc/rdb/impl/informix/InformixDc.java:217]: Indentation issue.

Comment (debnathbikram) [inline rdb/src/main/java/com/instana/dc/rdb/impl/informix/InformixDc.java:217]: Indentation issue.

File: rdb/src/main/java/com/instana/dc/rdb/impl/informix/InformixDc.java [hunk 3]
```diff
@@ -193,11 +208,14 @@ private void scheduleCustomPollRate(int pollInterval, IntervalType intervalType)
     private void parseCustomAttributes(Map<String, Object> properties) {
         Map<String, Object> customInput = (Map<String, Object>) properties.get("custom.input");
         String[] dbNames = ((String) customInput.get("db.names")).split(",");
+        int sequentialScanCount = (Integer)customInput.get("db.sequential.scan.count");
```
Comment (debnathbikram) [inline rdb/src/main/java/com/instana/dc/rdb/impl/informix/InformixDc.java:211]: Missing attribute in thfile config-informix.yaml

File: rdb/config/config-informix.yaml [hunk 4]
```diff
@@ -27,4 +27,6 @@ instances:
     #Custom Input
     custom.input:
       #Time Frame for which you want to get the expensive query. Default value 900 Sec
-      db.sql.elapsed.timeframe: 900   # Values should be in Seconds
\ No newline at end of file
+      db.sql.elapsed.timeframe: 900   # Values should be in Seconds
+      db.names: 'instana'
```
Comment (debnathbikram) [inline rdb/config/config-informix.yaml:31]: this is not required.

File: rdb/src/main/java/com/instana/dc/rdb/impl/informix/InformixDc.java [hunk 5]
```diff
@@ -197,6 +215,19 @@ private void scheduleCustomPollRate(int pollInterval, IntervalType intervalType)
...
     private void parseCustomAttributes(Map<String, Object> properties) {
+        Map<String, Object> customInput = (Map<String, Object>) properties.get("custom.input");
+        String[] dbNames = ((String) customInput.get("db.names")).split(",");
+        int sequentialScanCount = (Integer)customInput.get("db.sequential.scan.count");
+        StringBuilder sb = new StringBuilder(Constants.SINGLE_QUOTES + dbNames[0] + Constants.SINGLE_QUOTES);
+        for (int i = 1; i < dbNames.length; i++) {
+            sb.append(Constants.COMMA).append(Constants.SINGLE_QUOTES).append(dbNames[i].trim()).append(Constants.SINGLE_QUOTES);
+        }
+        tableSpaceSizeQuery = String.format(InformixUtil.TABLESPACE_SIZE_SQL, sb);
+        sequentialScanQuery = String.format(InformixUtil.DB_SEQ_SCAN_SQL,sb,sequentialScanCount);
```
Comment (debnathbikram) [inline rdb/src/main/java/com/instana/dc/rdb/impl/informix/InformixDc.java:226]: Indentation issue.

File: rdb/src/main/java/com/instana/dc/rdb/impl/informix/InformixDc.java [hunk 6]
```diff
@@ -197,6 +215,19 @@ private void scheduleCustomPollRate(int pollInterval, IntervalType intervalType)
...
+        sequentialScanTableQuery = String.format(InformixUtil.DB_SEQ_SCAN_TABLE_SQL,sb,sequentialScanCount);
```
Comment (debnathbikram) [inline rdb/src/main/java/com/instana/dc/rdb/impl/informix/InformixDc.java:227]: Indentation issue.

File: rdb/src/main/java/com/instana/dc/rdb/impl/informix/InformixUtil.java [hunk 7]
```diff
@@ -47,6 +60,7 @@ private InformixUtil() {
     //Disk Read & Write
     public static final String DB_DISK_WRITE_COUNT_SQL = "SELECT VALUE FROM SYSPROFILE WHERE NAME = 'dskwrites';";
     public static final String DB_DISK_READ_COUNT_SQL = "SELECT VALUE FROM SYSPROFILE WHERE NAME = 'dskreads';";
+>>>>>>> ab6c126af0988e786ae986bd3692850ab47f07b8
```
Comment (debnathbikram) [inline rdb/src/main/java/com/instana/dc/rdb/impl/informix/InformixUtil.java:63]: Remove this Merge conflicts changes.

File: rdb/src/main/java/com/instana/dc/rdb/impl/informix/InformixUtil.java [hunk 8]
```diff
@@ -38,6 +38,19 @@ private InformixUtil() {
...
+    public static final String DB_DATABASE_NLS_ENABLED_SQL = "SELECT is_nls, name as database_name FROM sysdatabases";
+    public static final String DB_DATABASE_CASE_INCENSITIVE_SQL = "SELECT is_case_insens, name as database_name FROM sysdatabases";
+
+    public static final String LOCK_OVF_SQL = "select coalesce(cast(value as int)) as value from sysprofile where name = 'ovlock'";
+    public static final String TRANSACTION_OVF_SQL = "select coalesce(cast(value as int)) as value from sysprofile where name = 'ovtrans'";
+    public static final String USER_OVF_SQL = "select coalesce(cast(value as int)) as value from sysprofile where name = 'ovuser'";
+    public static final String DB_SEQ_SCAN_SQL = "select coalesce(cast(seqscans as int))DATABASE_LOCK_TABLE_OVERFLOW,tabname as table_name from sysmaster:sysptprof where dbsname = %s and seqscans >= %d and tabname not like 'sys%%'";
+    public static final String DB_SEQ_SCAN_TABLE_SQL = "SELECT count(tabname)  as number_of_tables_having_sequential_scans FROM SYSPTPROF WHERE dbsname = %s and seqscans >= %d and tabname not like 'sys%%';";
+=======
```
Comment (debnathbikram) [inline rdb/src/main/java/com/instana/dc/rdb/impl/informix/InformixUtil.java:53]: Remove this Merge conflicts changes.

File: rdb/src/main/java/com/instana/dc/rdb/impl/informix/InformixDc.java [hunk 9]
```diff
@@ -197,7 +215,19 @@ private void scheduleCustomPollRate(int pollInterval, IntervalType intervalType)
...
-        Map<String, Object> customInput = (Map<String, Object>) properties.getOrDefault("custom.input", Collections.emptyMap());
+        Map<String, Object> customInput = (Map<String, Object>) properties.get("custom.input");
+        String[] dbNames = ((String) customInput.get("db.names")).split(",");
+        int sequentialScanCount = (Integer)customInput.get("db.sequential.scan.count");
+        StringBuilder sb = new StringBuilder(Constants.SINGLE_QUOTES + dbNames[0] + Constants.SINGLE_QUOTES);
+        for (int i = 1; i < dbNames.length; i++) {
+            sb.append(Constants.COMMA).append(Constants.SINGLE_QUOTES).append(dbNames[i].trim()).append(Constants.SINGLE_QUOTES);
+        }
+        tableSpaceSizeQuery = String.format(InformixUtil.TABLESPACE_SIZE_SQL, sb);
+        sequentialScanQuery = String.format(InformixUtil.DB_SEQ_SCAN_SQL,sb,sequentialScanCount);
```
Comment (debnathbikram) [inline rdb/src/main/java/com/instana/dc/rdb/impl/informix/InformixDc.java:226]:
```suggestion
        sequentialScanQuery = String.format(InformixUtil.DB_SEQ_SCAN_SQL, sb, sequentialScanCount);
```

File: rdb/src/main/java/com/instana/dc/rdb/impl/informix/InformixDc.java [hunk 10]
```diff
@@ -197,7 +215,19 @@ private void scheduleCustomPollRate(int pollInterval, IntervalType intervalType)
...
+        sequentialScanTableQuery = String.format(InformixUtil.DB_SEQ_SCAN_TABLE_SQL,sb,sequentialScanCount);
```
Comment (debnathbikram) [inline rdb/src/main/java/com/instana/dc/rdb/impl/informix/InformixDc.java:227]:
```suggestion
        sequentialScanTableQuery = String.format(InformixUtil.DB_SEQ_SCAN_TABLE_SQL, sb, sequentialScanCount);
```

File: rdb/src/main/java/com/instana/dc/rdb/impl/informix/InformixDc.java [hunk 11]
```diff
@@ -282,14 +317,12 @@ private void longPollingInterval() {
...
         getRawMetric(DB_TABLESPACE_MAX_NAME).setValue((List<SimpleQueryResult>) metricCollector.collectMetrics(DB_TABLESPACE_MAX_NAME));
-
         getRawMetric(DB_DATABASE_LOG_ENABLED_NAME).setValue((List<SimpleQueryResult>) metricCollector.collectMetrics(DB_DATABASE_LOG_ENABLED_NAME));
         getRawMetric(DB_DATABASE_BUFF_LOG_ENABLED_NAME).setValue((List<SimpleQueryResult>) metricCollector.collectMetrics(DB_DATABASE_BUFF_LOG_ENABLED_NAME));
         getRawMetric(DB_DATABASE_ANSI_COMPLAINT_NAME).setValue((List<SimpleQueryResult>) metricCollector.collectMetrics(DB_DATABASE_ANSI_COMPLAINT_NAME));
         getRawMetric(DB_DATABASE_NLS_ENABLED_NAME).setValue((List<SimpleQueryResult>) metricCollector.collectMetrics(DB_DATABASE_NLS_ENABLED_NAME));
         getRawMetric(DB_DATABASE_CASE_INCENSITIVE_NAME).setValue((List<SimpleQueryResult>) metricCollector.collectMetrics(DB_DATABASE_CASE_INCENSITIVE_NAME));
-
-    }
+           }
```
Comment (debnathbikram) [inline rdb/src/main/java/com/instana/dc/rdb/impl/informix/InformixDc.java:325]: Indentation is not right.

File: rdb/src/main/java/com/instana/dc/rdb/impl/informix/InformixUtil.java [hunk 12]
```diff
@@ -38,11 +38,18 @@ private InformixUtil() {
...
+    public static final String DB_DATABASE_BUFF_LOG_ENABLED_SQL = "SELECT is_buff_log, name as database_name FROM sysdatabases";
+    public static final String DB_DATABASE_ANSI_COMPLAINT_SQL = "SELECT is_ansi, name as database_name FROM sysdatabases";
+    public static final String DB_DATABASE_NLS_ENABLED_SQL = "SELECT is_nls, name as database_name FROM sysdatabases";
+    public static final String DB_DATABASE_CASE_INCENSITIVE_SQL = "SELECT is_case_insens, name as database_name FROM sysdatabases";
+
+    public static final String LOCK_OVF_SQL = "select coalesce(cast(value as int)) as value from sysprofile where name = 'ovlock'";
+    public static final String TRANSACTION_OVF_SQL = "select coalesce(cast(value as int)) as value from sysprofile where name = 'ovtrans'";
+    public static final String USER_OVF_SQL = "select coalesce(cast(value as int)) as value from sysprofile where name = 'ovuser'";
+    public static final String DB_SEQ_SCAN_SQL = "select coalesce(cast(seqscans as int))DATABASE_LOCK_TABLE_OVERFLOW,tabname as table_name from sysmaster:sysptprof where dbsname = %s and seqscans >= %d and tabname not like 'sys%%'";
```
Comment (debnathbikram) [inline rdb/src/main/java/com/instana/dc/rdb/impl/informix/InformixUtil.java:51]: What is the purpose of two `%`  in the `'sys%%'` ?

File: rdb/src/main/java/com/instana/dc/rdb/impl/informix/InformixUtil.java [hunk 13]
```diff
@@ -38,11 +38,18 @@ private InformixUtil() {
...
+    public static final String DB_SEQ_SCAN_TABLE_SQL = "SELECT count(tabname)  as number_of_tables_having_sequential_scans FROM SYSPTPROF WHERE dbsname = %s and seqscans >= %d and tabname not like 'sys%%';";
```
Comment (debnathbikram) [inline rdb/src/main/java/com/instana/dc/rdb/impl/informix/InformixUtil.java:52]: What is the purpose of two `%`  in the `'sys%%'` ?

Comment (vivek-22) [reply rdb/src/main/java/com/instana/dc/rdb/impl/informix/InformixUtil.java:52]: In the query , we need to escape the '%' with another '%' character. Else, its throwing error. Hence multiple '%%'.

File: rdb/src/main/java/com/instana/dc/rdb/impl/informix/InformixUtil.java (hunk 12 above)
Comment (vivek-22) [reply rdb/src/main/java/com/instana/dc/rdb/impl/informix/InformixUtil.java:51]: In the query , we need to escape the '%' with another '%' character. Else, its throwing error. Hence multiple '%%'.

File: rdb/src/main/java/com/instana/dc/rdb/impl/informix/InformixDc.java [hunk 14]
```diff
@@ -197,9 +215,12 @@ private void scheduleCustomPollRate(int pollInterval, IntervalType intervalType)
      */
     @SuppressWarnings("unchecked")
     private void parseCustomAttributes(Map<String, Object> properties) {
-        Map<String, Object> customInput = (Map<String, Object>) properties.getOrDefault("custom.input", Collections.emptyMap());
+        Map<String, Object> customInput = (Map<String, Object>) properties.get("custom.input");
+        int sequentialScanCount = (Integer)customInput.get("db.sequential.scan.count");
         long elapsedTimeFrame = Long.parseLong((customInput.getOrDefault("db.sql.elapsed.timeframe", DEFAULT_ELAPSED_TIME)).toString());
         StringBuilder databaseName = new StringBuilder(Constants.SINGLE_QUOTES + getDbName() + Constants.SINGLE_QUOTES);
+        sequentialScanQuery = String.format(InformixUtil.DB_SEQ_SCAN_SQL,databaseName,sequentialScanCount);
```
Comment (debnathbikram) [inline rdb/src/main/java/com/instana/dc/rdb/impl/informix/InformixDc.java:222]:
```suggestion
        sequentialScanQuery = String.format(InformixUtil.DB_SEQ_SCAN_SQL, databaseName, sequentialScanCount);
```

**Review by debnathbikram — APPROVED** [review]
LGTM

**Pull request merged**